Add proposed feature specification for "implicit names". (#4538)
dart-lang/language: working/3102 - implied-name/feature-specification.md

    return (:red, :green, :blue, :alpha);
  }
}
```

## Specification

The current grammar for a named argument is:
```ebnf
<namedArgument> ::= <label> <expression>
```
The [current grammar](../../accepted/3.0/records/feature-specification.md#record-expressions) for a record literal field:
```ebnf
recordField  ::= (identifier ':' )? expression
```
So basically the same.

The new grammar is:
```ebnf
<namedArgument> ::= <namedExpression>

<recordField> ::= <namedExpression> | <expression>

<namedExpression> ::= <identifier>? `:' <expression>
```

This grammar change does not introduce any new ambiguities.
A named record field and a named argument can *only* occur
immediately after a `(` or a `,`, as the entire content until
a following `,` or `)`.
Starting with a `:` at that point is not currently possible.
(Even if we allow metadata annotations inside argument lists or record literals,
it's unambiguous whether those annotations includes a following identifier or
not.)

As a non-grammatical restriction, it's a **compile-time error**
if a `<namedExpression>` omits the leading`<identifier>`, and the following
expressions is not a _single identifier expression_, as defined below.

> In short, an expression is a _single identifier expression_

Reviewer: ```suggestion
An expression is a _single identifier expression with identifier_
```

which has a specific single _identifier_ if and only if
it is one of the following:
*   An identifier, with that identifier.

> *   `s!`, `s as T`, `(s)`, `s..cascadeSection`/`s?..cascadeSection`

Author: Didn't find any examples of a `id: id?..selector..selector` in the wild.
I added examples of using `?..`, but not two selectors.

    where `s` is a single identifier expression,
    and then it has the same identifier as `s`.

More formally, An expression `e` is a single identifier expression with a certain identifier if and only if it is defined as such by the following:

*   If `e` is a `<primary>` expression which is an `<identifier>`,
    it is a single identifier expression with that `<identifier>` as identifier.
*   `s!`: If `e` is  a `<primary> <selector>*` production where `<selector>*`
    is the single `<selector>` `` `!' ``, and `<primary>` is
    a single identifier expression, then `e` is a single identifier expression

>     with the same identifier as the `<primary>`,

Reviewer 2: In the interest of completeness, I'm compelled to ask about `(:foo++)` and `(:foo--)` as well.

(I could go either way with them. But I'd lean towards "no".)

Author: I'd argue against `-foo`, `!foo` and `~foo` in that they *don't* mean the same thing as the name. They likely mean the _exact opposite_.

Doing `foo(isOpen: !isOpen)` or `foo(count: -count)` would look wrong, and allowing `foo(:!isOpen)` and `foo(:-count)` would hide that.

The only operators allowed are ones that do not change the value of the identifier expression, they just might throw before evaluating to that value (`!` or `as Type`).

For `id++` and `++id`, the value of the expression *is* the value of the variable, either before or after the variable is incremented, that's why they could be argued to fit. But as I said, a mutable variable is mutable state, being the state can mean something else than just its value. This feature is _mainly_ intended for
_forwarding values_ from one place to another with the same name.

*   `(s)`: If `e` is a `<primary>` production of the form
    `` `(' <expression> `)' `` and the `<expression>`
    is a single-identifier expression,
    then `e` is a single identifier expression with the same identifier as
    the `<expression>`.
*   `s..cascade`:
    *   If `e` a `<cascade>` of the form ``<cascade> `..' <cascadeSection>``
        and the `<cascade>` is a single identifier expression,
        then `e` is a single identifier expression with the same identifier as
        the `<cascade>`.
    *   If `e` a `<cascade>` of the form
        ``<conditionalExpression> (`?..' | `..') <cascadeSection>``
        and the `<conditionalExpression>` is a single identifier expression,
        then `e` is a single identifier expression with the same identifier as
        the `<conditionalExpression>`.

The _name of a named expression_ is then:
* If the named expression has a leading identifier, then that identifier.
* Otherwise the following expression must be a single identifier expression
  with an identifier *I*, and then the name of the named expression is *I*.

The name of a `<namedArgument>` or a named `<recordField>` is the name of
its `<namedExpression>`.

> 

Reviewer: Typo:

```suggestion
```
